### ISSUE TITLE
Implement Alt-O to Swap Spec/Body Similar to Existing VSCode C++ Feature

### ISSUE THREAD

**Issue by cwilliso813** [AdaCore/ada_language_server#669]
With the Microsoft C++ extension, the Alt-O shortcut can be used to swap between editing the header and source file for a particular header/source pair with the same name.

This would be a very useful feature for Ada since it has the corresponding spec/body pairing, and developers often have to swap between these two files.

There are several naming conventions in play for Ada: (*.ada, *.ads/*.adb, *.1.ada/*.2.ada) and the extension would have to handle this automatically or provide some sort of configurability for naming convention.

**Comment by 1ma** [comment]
FYI there is already a similar feature that I use all the time: Ctrl+Clicking the name of a function in the Spec file takes you to the implementation in the Body file, and vice-versa.

**Comment by cwilliso813** [comment]
That is helpful but I still think Alt-O would add value.

In addition, I regret to say that many of the features of the VSCode Ada extension do not work for me or work inconsistently, so I am unable to test this properly. My best guess on why the extension does not work is that my codebase is being accessed by necessity through sshfs which is resulting in network bottlenecks. I often see VSCode doing long-running indexing operations on the filesystem and when the extension does work, navigation is painfully slow (several seconds of holding control before Click can be performed).

Separate issue but perhaps my problems would be solved if the extension could be optimized for sshfs.

**Comment by AnthonyLeonardoGracio** [comment]
Hello!

> That is helpful but I still think Alt-O would add value.

Unfortunately the LSP does not provide any request to provide a 'Switch to header/body' command. But here is an extension that works: https://marketplace.visualstudio.com/items?itemName=ryzngard.vscode-header-source

You will just need to add an entry for Ada in the `headerSourceSwitch.mappings` setting to have Alt+O working for Ada too.

> In addition, I regret to say that many of the features of the VSCode Ada extension do not work for me or work inconsistently, so I am unable to test this properly. My best guess on why the extension does not work is that my codebase is being accessed by necessity through sshfs which is resulting in network bottlenecks. I often see VSCode doing long-running indexing operations on the filesystem and when the extension does work, navigation is painfully slow (several seconds of holding control before Click can be performed).
> 
> Separate issue but perhaps my problems would be solved if the extension could be optimized for sshfs.

Indeed, looks like a separate issue. Looking a bit on the web, it seems that some people are recommending the following extension for VS Code: https://github.com/liximomo/vscode-sftp

Maybe worth to give it a try ;) In any case, we can't do much on the Ada Language Server side.

Regards,

**Comment by reznikmm** [comment]
@cwilliso813 Speaking about `sshfs`, I would suggest you to try Remote Development extension for VS Code instead. I've tried it with the Ada extension and it works well!

I made a short video about it:

[![](http://img.youtube.com/vi/VjS7L_rQY2A/0.jpg)](http://www.youtube.com/watch?v=VjS7L_rQY2A "Remote Development in Ada")

**Comment by cwilliso813** [comment]
Thanks for the suggestion! I tried the remote extension with SSH and I think it should be a major improvement over using VSCode overtop sshfs, which tends to be unstable in my environment. I also managed to install the Ada Language Support extension on the remote host using this method. The install was a little rough because the remote machine lacks at internet connection. Setting "remote.SSH.localServerDownload" to "always" seemed to help. What I wanted was for the extension to be downloaded on the client and transferred to the host. That took several tries and restarts of VSCode, but eventually took.

I then open a folder on the remote machine and tried Go To Definition, Find All References, Ctrl + Click, etc. Interesting findings:
- Sometimes navigation works
- Performance is often very slow (upwards of 1 minute of "Loading..." or a horizontal progress bar before navigation)
- In some files navigation appears to "give up" and displays "No Definition Found" and similar

Based on what I've seen, I'd say the issue is not entirely related to using sshfs as I'd thought. This is a large codebase. The Ada Language Extension may be having trouble with the scale. I feel like it could work if there was some way to make a few tweaks to its searching and caching.
